compiler/rustc_target/src/spec/mod.rs

```diff
@@ -2433,9 +2446,21 @@ impl DerefMut for Target {
 
 impl Target {
     /// Given a function ABI, turn it into the correct ABI for this target.
-    pub fn adjust_abi(&self, abi: Abi) -> Abi {
+    pub fn adjust_abi<C>(&self, cx: &C, abi: Abi) -> Abi
+    where
+        C: HasWasmCAbiOpt,
+    {
         match abi {
-            Abi::C { .. } => self.default_adjusted_cabi.unwrap_or(abi),
+            Abi::C { .. } => {
+                if self.arch == "wasm32"
+                    && self.os == "unknown"
+                    && cx.wasm_c_abi_opt() == WasmCAbi::Spec
+                {
+                    abi
+                } else {
+                    self.default_adjusted_cabi.unwrap_or(abi)
+                }
```

Comment:
At this point, wouldn't it be cleaner to remove `default_adjusted_cabi` entirely, and make this logic here more like
```rust
if this_is_wasm32_unknown() && cx.wasm_c_abi_opt() == Legacy {
  Abi::Wasm 
} else {
  abi
}
```

Currently there's 2 level of adjustments overriding each other and that's rather hacky IMO.

Reply:
I don't disagree, but seeing that we can't remove `default_adjusted_cabi` because it's used by other targets, we would make it no-op, which is a bit weird to me. E.g. we would overwrite or ignore whatever would be set in `default_adjusted_cabi`.

If I remember correctly, this could be simplified by making the flag change the instantiated targets `default_adjusted_cabi`, I've seen other fields adjusted but not through flags, which is why I didn't do this in the first place.

Will look into it again.

Reply:
I was saying this under the assumption that `default_adjusted_cabi` is only used by this target. That's what I concluded after a quick grep. But maybe I missed something. Which other target is using `default_adjusted_cabi`?

Reply:
Uhm, you are right, it seems to only be used by Wasm. From the top of my head I seem to remember it being used by the CUDA target, either it was removed since last time or I just remember it wrongly.

Will remove it.

Reply:
Done.
Thank you!